design/api.md

```diff
@@ -0,0 +1,219 @@
+# Fission Server API Specification
+
+## Abstract
+
+This spec describes
+- available [server routes](#routes) and which [UCAN] capabilities the these routes expect
+- the [set of capabilities](#ucan-capabilities) the server understands and how they can imply each other.
+
+
+## Routes
+
+Today we use a REST API. In the future we may switch to an RPC-based design. When/if this happens mostly depends on a plan for implementing the UCAN invocation spec.
+
+### POST `/api/v0/auth/email/verify`
+
+---
+
+**Authentication**: *Unauthenticated*
+
+**Request**: `{ email: string }`
+
+**Response**: `{ success: bool }`
+
+---
+
+- Sends an email with a 6-digit verification code to given address
+- A hash of the code and email address is stored in a database
+- The code expires after 24 hours
+
+
+### POST `/api/v0/account`
+
+---
+
+**Authentication**:
+
+UCAN giving access to the capability
+- with resource `fission:*`
+- with ability `account/create` (without caveats)
+- originating in `did:key:<device>`.
+Its audience must be `did:<server DID>`.
+
+**Request**:
+```ts
+{
+  code: number,
+  email: string,
+  username: string,
+  did: "did:key:<device>",
+  credentialID?: string,
+}
+```
+
+**Response**:
+
+```ts
+{ ucans: Array<string> }
+```
+
+The response consists of two UCANs:
+- ```json
+  {
+    "iss": "did:key:<account>",
+    "aud": "did:<server DID>",
+    "cap": { "fission:did:key:<account>": { "*": [{}] } }
+  }
+  ```
+- ```json
+  {
+    "iss": "did:<server DID>",
+    "aud": "did:key:<device>",
+    "cap": { "fission:did:key:<account>": { "*": [{}] } }
+  }
+  ```
+
+---
+
+This request registers a full account given an email verification code and delegates the account's access to given DID. In the future more UCANs may be involved in the delgation chain.
+
+
+### GET `/api/v0/account/:did`
+
+---
+
+**Authentication**:
+
+UCAN giving access to the capability `"fission:did:key:<account>": { "account/info": [{}] }` originating in `did:key:<account>`.
+The UCAN's audience needs to be `did:key:<account>`.
+
+**Response**:
+
+```ts
+{
+  email: string,
+  did: string,
+  username: string,
+}
```

Comment:
Not sure if helpfuk here, but in the UCAN spec, I've started putting these in tables because I can stick comments right next to each field.

| Field | Type | Format | Required | Comment |
|-----|-------|---------|---|--------|
| `email` | `string` | Email | Yes | User email |
| `did` | `string` | DID | Yes | The DID to bind the email to |
| `username` | `string` | UTF8 | Yes | The name that the user wants to register |

You could also do this with TS I guess you could also do this with TS comments, but

Reply:
Tables would be nicer, agreed. I started out putting *all of the things* into tables first, but abandoned because
- it took me too long
- multiline cells suck.

But if I keep it only to records like this both points probably vanish. It definitely looks nicer, when I do another pass, I'll change it :)